File: pyalgorithm/array/container_with_most_water.py
from typing import List

import unittest


class Solution(unittest.TestCase):

    def maxArea(self, nums: List[int]) -> bool:
        area_max = 0
        left, right = 0, len(nums) - 1
        while left < right:
            width = right - left
            heigh = min(nums[left], nums[right])
            area = width * heigh
            area_max = max(area, area_max)
            if nums[left] < nums[right]:
                left += 1
            elif nums[left] > nums[right]:
                right -= 1
            else:
                left += 1
                right -= 1

        return area_max

    def test(self):
        self.assertEqual(49, self.maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]))


if __name__ == '__main__':
    unittest.main()

File: leetcode/src/main/java/leetcode/array/ContainerWithMostWater.java
package leetcode.array;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;


/**
 * Chủ đề: Two pointer
 *
 * https://leetcode.com/problems/container-with-most-water/
 *
 * Cho mảng chiều cao, tìm 2 điểm mà diện tích nước trong đó là nhiều nhất
 */
public class ContainerWithMostWater {

    public int maxArea(int[] arr) {
        int maxArea = 0;
        int left = 0, right = arr.length - 1;
        while (left < right) {
            int width = right - left; // Chiều rộng
            int hight = Math.min(arr[left], arr[right]); // Chiều cao
            int area = width * hight; // Tính diện tích
            maxArea = Math.max(maxArea, area); // Tìm diện tích lớn nhất
            // Dời con trỏ
            if (arr[left] < arr[right]) {
                left++;
            } else if (arr[left] > arr[right]) {
                right--;
            } else {
                left++;
                right--;
            }
        }
        return maxArea;
    }

    @Test
    public void test() {
        assertEquals(1, maxArea(new int[]{1, 1}));
        assertEquals(49, maxArea(new int[]{1, 8, 6, 2, 5, 4, 8, 3, 7}));
        assertEquals(24, maxArea(new int[]{1, 3, 2, 5, 25, 24, 5}));
        assertEquals(17, maxArea(new int[]{2, 3, 4, 5, 18, 17, 6}));
    }

}
